chat-android/src/main/java/com/ably/chat/Typing.kt

```diff
@@ -119,130 +134,234 @@ internal class DefaultTyping(
 
-    private val eventBus = MutableSharedFlow<Unit>(
-        extraBufferCapacity = 1,
-        onBufferOverflow = BufferOverflow.DROP_OLDEST,
-    )
+    private var typingHeartBeatStarted: Long? = null
 
     override val channelWrapper: RealtimeChannel = room.realtimeClient.channels.get(typingIndicatorsChannelName, ChatChannelOptions())
 
     @OptIn(InternalAPI::class)
     override val channel: Channel = channelWrapper.javaChannel // CHA-RC2f
 
-    private var typingJob: Job? = null
-
     private val listeners: MutableList<Typing.Listener> = CopyOnWriteArrayList()
 
-    private var lastTyping: Set<String> = setOf()
+    private var currentlyTypingMembers: MutableSet<String> = mutableSetOf()
+
+    private var typingEventPubSubSubscription: Subscription
+
+    /**
+     * A mutable map of clientId to TimedTypingStopEvent job.
+     * Each received typing start event is capable of timed death after a certain period (heartbeatThrottle + timeout).
+     */
+    private val typingStartEventPrunerJobs = mutableMapOf<String, Job>()
+
+    /**
+     * Defines how often a typing.started heartbeat can be sent (in milliseconds).
+     * This does not prevent a client from sending an typing.stopped heartbeat—doing so resets the interval timer
+     * and allows another typing.started to be sent.
+     * Spec: CHA-T10
+     */
+    private val heartbeatThrottle: Duration = room.options.typing?.heartbeatThrottle ?: throw AblyException.fromErrorInfo(
+        ErrorInfo(
+            "Typing options hasn't been initialized",
+            ErrorCode.BadRequest.code,
+        ),
+    )
+
+    /**
+     * Defines how long to wait before marking a user as stopped typing.
+     * For example, if the last typing.started heartbeat was received 10s ago, the system will wait an additional 2s before declaring
+     * the client as no longer typing.
+     * @defaultValue 2 seconds
+     * Spec: CHA-T10a
+     */
+    private val timeout: Duration = 2.seconds
 
-    private var presenceSubscription: Subscription
+    private val eventBus = MutableSharedFlow<SelfTypingEvent>(extraBufferCapacity = kotlinx.coroutines.channels.Channel.UNLIMITED)
 
+    /**
+     * Spec: CHA-T13
+     */
     init {
+        // Send self typing events
         typingScope.launch {
-            eventBus.collect {
-                processEvent()
-            }
+            eventBus.collect { sendSelfTypingEvent(it) }
         }
 
-        val presenceListener = PresenceListener {
-            if (it.clientId == null) {
-                logger.error("unable to handle typing event; no clientId", context = mapOf("member" to it.toString()))
-            } else {
-                eventBus.tryEmit(Unit)
+        // Process received typing events
+        val typingListener = PubSubMessageListener { msg ->
+            val typingEventType = TypingEventType.entries.first { it.eventName == msg.name }
+            // CHA-T13a
+            if (msg.data == null || msg.data !is String) {
```

Comment:
Bit confused here - we're checking `msg.data` but then logging an error to do with clientId?

Reply:
In any case, at the moment we don't care about `msg.data` for typing - just `msg.name`

Reply:
Okay, I just checked the code and it was intentional not to put `msg.name` check here. 
If you see, we are already registering listeners specifically for typing start and stop using 
```
       val typingEvents = listOf(TypingEventType.Started.eventName, TypingEventType.Stopped.eventName)
```
Also, in the pubsub SDKs, we call those listeners using `msg.name`, so there won't be a case when `msg.name` will ever be malformed. 
Link to the code -> https://github.com/ably/ably-java/blob/64575b867484d2ef89804d066b390228041c8713/lib/src/main/java/io/ably/lib/realtime/ChannelBase.java#L897

Reply:
we should check `clientId` here, no need to put `clientId` in the message data